### PR TITLE
LibGfx+LibWeb: Use harfbuzz for text shaping

### PR DESCRIPTION
**Pull request by BenJilks**
This is further progress on issue https://github.com/LadybirdBrowser/ladybird/issues/862. Characters are now rendered in the correct order, however text chunks are still ordered left to right.

This does add the extra dependency of harfbuzz. It was already included a part of skia, it’s just marked explicitly as a dependency now. It is a C library, so doesn’t integrate super well with the memory model. So I’ve tied the harfbuzz objects to the lifetime of their respective LibGfx objects.

### CODE REVIEW

File: Userland/Libraries/LibGfx/TextLayout.cpp [hunk 1]
```diff
@@ -7,51 +7,85 @@
...
-    if (font_contains_glyph && !check_for_emoji) {
-        return DrawGlyph {
-            .position = point,
-            .code_point = code_point,
-        };
+    u32 glyph_count;
+    auto* glyph_info = hb_buffer_get_glyph_infos(buffer, &glyph_count);
+    Vector<hb_glyph_info_t> const input_glyph_info({ glyph_info, glyph_count });
+    if (input_glyph_info.is_empty()) {
+        hb_buffer_destroy(buffer);
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibGfx/TextLayout.cpp:60]: nit: using AK::ScopeGuard at the top of a scope might a better idea.

File: Userland/Libraries/LibGfx/TextLayout.cpp [hunk 2]
```diff
@@ -7,51 +7,85 @@
...
+    hb_shape(hb_font, buffer, nullptr, 0);
+
+    glyph_info = hb_buffer_get_glyph_infos(buffer, &glyph_count);
+    auto* positions = hb_buffer_get_glyph_positions(buffer, &glyph_count);
+
+    FloatPoint point = baseline_start;
+    for (size_t i = 0; i < glyph_count; ++i) {
+        auto position = point
+            - FloatPoint { 0, font.pixel_metrics().ascent }
+            + FloatPoint { positions[i].x_offset, positions[i].y_offset } / 64.0f;
```
[the remark this answers is not in the document]
Comment (BenJilks) [reply Userland/Libraries/LibGfx/TextLayout.cpp]: Thanks for reviewing. I've made that a constant in `Font.h`.

File: Userland/Libraries/LibGfx/TextLayout.cpp (hunk 1 above)
Comment (BenJilks) [reply Userland/Libraries/LibGfx/TextLayout.cpp]: It's using a scope guard now.

**Review by kalenikaliaksandr — CHANGES_REQUESTED** [review]
The commit touches both LibWeb and LibGfx, so it should be:
"LibGfx+LibWeb: Use harfbuzz for text shaping"

**Pull request merged**